packages/vaex-ml/vaex/ml/transformations.py

```diff
@@ -121,6 +121,117 @@ def transform(self, df, n_components=None):
+class RandomProjections(Transformer):
+    '''Reduce dimensionality through a random matrix projection.
+
+    The random projections method is based on the Johnson-Lindenstrauss lemma.
+    For mode details see https://en.wikipedia.org/wiki/Johnson%E2%80%93Lindenstrauss_lemma
+
+    Note that you need scikit-learn to fit this Transformer but not for transformations using an already fitter Transformer.
+
+    Example:
+
+    >>> import vaex
+    >>> import vaex.ml
+    >>> df = vaex.from_arrays(x=[2,5,7,2,15], y=[-2,3,0,0,10], z=[2, -10, 2, 3, 0])
+    >>> df
+    #    x    y    z
+    0    2   -2    2
+    1    5    3  -10
+    2    7    0    2
+    3    2    0    3
+    4   15   10    0
+    >>> rand_proj = vaex.ml.RandomProjections(features=['x', 'y', 'z'], n_components=2)
+    >>> rand_proj.fit_transform(df)
+    #    x    y    z    random_projection_0    random_projection_1
+    0    2   -2    2                1.73363             -0.0700273
+    1    5    3  -10              -17.8742             -14.0226
+    2    7    0    2               -3.32911             -8.50181
+    3    2    0    3                2.04843             -1.27538
+    4   15   10    0              -17.0289             -28.6562
+    '''
+    snake_name = 'random_projections'
+    n_components = traitlets.CInt(default_value=None, allow_none=True, help='Number of components to retain. If None (default) the number will be set via the Johnson-Lindenstrauss formula. See https://scikit-learn.org/stable/modules/generated/sklearn.random_projection.johnson_lindenstrauss_min_dim.html for more details.')
+    eps = traitlets.Float(default_value=0.1, allow_none=True, help='Parameter to control the quality of the embedding according to the Johnson-Lindenstrauss lemma when `n_components` is set to None. The value must be positive.')
+    matrix_type = traitlets.Enum(values=['gaussian', 'sparse'], default_value='gaussian', help='The type of random matrix to create. The values can be "gaussian" and "sparse".')
+    density = traitlets.Float(default_value=None, allow_none=True, help='Ratio in the range (0, 1] of non-zero component in the random projection matrix. Only valid if `matrix_type` is "sparse". If density is None, the value is set to the minimum density as recommended by Ping Li et al.: 1 / sqrt(n_features).')
+    prefix = traitlets.Unicode(default_value="random_projection_", help=help_prefix)
+    random_state = traitlets.Int(default_value=None, allow_none=True, help='Controls the pseudo random number generator used to generate the projection matrix at fit time. Used to get reproducible results.')
+    random_matrix_ = traitlets.List(traitlets.List(traitlets.CFloat()), help='The random matrix.').tag(output=True)
+
+    @traitlets.validate('eps')
+    def _valid_eps(self, proposal):
+        if (proposal['value'] > 0) & (proposal['value'] < 1):
+            return proposal['value']
+        else:
+            raise traitlets.TraitError('`eps` must be between 0 and 1.')
```

Comment:
we can also rely on min/max of the trait, like traitlets.Int(..., min=0, max=1), however, those bounds are inclusive.

Reply:
Yeah i tried that and found that out. Do you mind leaving it like this, or do you propose a chance of some kind?

Reply:
no, if this was the reason, we keep it like this!